### PR TITLE
Collect resource timing and other analytics metrics

### PR DESCRIPTION
**Pull request by querymetrics**
Related to https://github.com/ampproject/amphtml/issues/3022

This new extension (`amp-analytics-resources`) adds extra variables to the amp-analytics and amp-pixel URL substitutions.
The extension is optional and if not loaded the URL substitution for the extra variables should return empty strings.
- compressed ResourceTiming data
- # of resources referenced by the page
- # of different domains referenced by the page
- # of DOM nodes on the page
- byte size of rendered page
- # of image nodes on the page
- # of image nodes on the page that reference an external domain
- # of script nodes on the page
- # of script nodes on the page that reference an external domain

Still needs documentation.

cc @nicjansma

### CODE REVIEW

**Comment by avimehta** [comment]
Mostly lgtm from my side. Some comments:

- There was a presubmit failure. Please make sure `gulp check-types` passes.
- amp-analytics-resources.js can probably be merged with the -impl.js file. Any thoughts on why not to do it?
- @jridgewell Who should look at the 3p contribution?

**Comment by rudygalfi** [comment]
What's the rationale for doing a totally new extension? Is this set of functionality pretty heavyweight and best not exposed to everyone who implements amp-analytics?

**Comment by avimehta** [comment]
@rudygalfi See #3022 for details. The primary reason is the size of additional logic.

**Comment by querymetrics** [comment]
@avimehta , the `-impl.js` was split out so that we could control the call to `installAmpAnalyticsResourcesService()` in the unit tests with the correct `window` object.
It might allow us to make more unit tests in `test-url-replacements.js` that simulate the extension not being loaded as well.

**Comment by querymetrics** [comment]
We made some changes to the `resourcetiming-compression` library so that it wouldn't need to be forked/modified to be used in this extension. Also added ResourceTiming2 data (`encodedBodySize`, `decodedBodySize` and `transferSize`) for browsers that support it.

It's now ready for review.

**Comment by avimehta** [comment]
@querymetrics sorry for the delay. I am really swamped right now. Will review this PR in next week or so.

**Comment by avimehta** [comment]
lgtm. Can you add an example in the examples/ directory? 

/cc @dvoytenko for a second pair of eyes on a new extension.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+      result = nodes;
+    }
+    return result;
+  }
+}
+
+/**
+ * @param {!Window} win
+ * @return {!Object} All services need to return an object to "load".
+ * @visiblefortesting
```
Comment (avimehta) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: Not sure if @visiblefortesting is used in this repo much.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    const res = this.getResources_();
+    const doms = {};
+    let a;
+    if (!res || !this.win_['document']
+        || typeof this.win_['document']['createElement'] !== 'function') {
+      return Promise.resolve();
+    }
+
+    // use an 'a' element to do the url parsing
+    a = this.win_.document.createElement('a');
```
Comment (avimehta) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: use parseUrl from https://github.com/ampproject/amphtml/blob/master/src/url.js#L66?

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+  /**
+   * Returns number of distinct domains referenced from the page.
+   */
+  getDomainCount() {
+    const res = this.getResources_();
+    const doms = {};
+    let a;
+    if (!res || !this.win_['document']
+        || typeof this.win_['document']['createElement'] !== 'function') {
```
Comment (avimehta) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: || on previous line.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+/** @private Visible for testing. */
+export class AnalyticsResourcesService {
+  constructor(window) {
+    /** @const {!Window} */
+    this.win_ = window;
+  }
+
+  /**
+   * Rendered byte count of base page. This may not be what was sent on the network.
```
Comment (avimehta) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: here and below. Add @return {type} to the docs? you can still keep the single line docs if possible.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+      return [];
+    }
+    return p.getEntriesByType('resource');
+  }
+
+  /**
+   * Get DOM elements that match type and filter
+   * @param {String} element type
+   * @param {function(Object)=} filter
+   * @return {!Array} array of DOM elements
```
Comment (avimehta) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: !Array<Element>?

**Comment by querymetrics** [comment]
@avimehta, thanks for the comments. 
To help debug, I put up a very basic form that you can paste compressed resource timing strings into and see the decompressed data: http://querymetrics.com/rtd.html

**Comment by jridgewell** [comment]
Assigning to @dvoytenko for the third party code.

File: build-system/tasks/compile.js [hunk 6]
```diff
@@ -136,6 +136,7 @@ function compile(entryModuleFilename, outputDir,
       // Not sure what these files are, but they seem to duplicate code
       // one level below and confuse the compiler.
       '!node_modules/core-js/modules/library/**.js',
+      'node_modules/resourcetiming-compression/src/resourcetiming-compression.js',
```
Comment (dvoytenko) [inline build-system/tasks/compile.js]: /cc @erwinmombay PTAL this part.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    /** @private @const {!Window} */
+    this.win_ = win;
+  }
+
+  /**
+   * Returns rendered byte count of base page. This may not be what was sent on the network.
+   * @return {!Promise<string|undefined>}
+   */
+  getDocumentLength() {
+    const innerHTML = this.win_['document'] &&
```
Comment (dvoytenko) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: Let's just use a normal properties notation here.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+/** @private Visible for testing. */
+export class AnalyticsResourcesService {
+  constructor(win) {
+    /** @private @const {!Window} */
+    this.win_ = win;
+  }
+
+  /**
+   * Returns rendered byte count of base page. This may not be what was sent on the network.
+   * @return {!Promise<string|undefined>}
```
Comment (dvoytenko) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: Here and below: is there any reason we shouldn't just return the actual types? E.g. `number`?

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    return (res && typeof res.length === 'number') ?
+        Promise.resolve(String(res.length)) : Promise.resolve();
+  }
+
+  /**
+   * Returns number of DOM nodes on the page.
+   * @return {!Promise<string|undefined>}
+   */
+  getDomNodeCount() {
+    const nodes = this.getNodes_('*');
```
Comment (dvoytenko) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: Do you have any data on how slow/fast this is in a typical document?

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    return (nodes && typeof nodes.length === 'number') ?
+        Promise.resolve(String(nodes.length)) : Promise.resolve();
+  }
+
+  /**
+   * Returns number of image nodes on the page.
+   * @return {!Promise<string|undefined>}
+   */
+  getDomImgCount() {
+    const nodes = this.getNodes_('img');
```
Comment (dvoytenko) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: I'm wondering if this is what you actually want. In AMP `img` tags are created lazily as they are the user scrolls through the content. Thus, some `amp-img` might exist but not yet "loaded" their respective `img` tags.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+        Promise.resolve(JSON.stringify(rt)) : Promise.resolve();
+  }
+
+  /**
+   * Get ResourceTiming entries.
+   * @return {Array<PerformanceResourceTiming>}
+   * @private
+   */
+  getResources_() {
+    const p = ('performance' in this.win_) && this.win_['performance'];
```
Comment (dvoytenko) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: Ditto everywhere on property access.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+      return null;
+    }
+    return p.getEntriesByType('resource');
+  }
+
+  /**
+   * Get DOM elements that match type and filter.
+   * @param {String} element type
+   * @param {function(Object)=} filter
+   * @return {Array<HTMLElement>} DOM elements
```
Comment (dvoytenko) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: It looks like results are always `!Array<!Element>`.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+      // ResourceTiming interface is not supported.
+      return null;
+    }
+    return p.getEntriesByType('resource');
+  }
+
+  /**
+   * Get DOM elements that match type and filter.
+   * @param {String} element type
+   * @param {function(Object)=} filter
```
Comment (dvoytenko) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: Please rename to `opt_filter`.

File: extensions/amp-analytics-resources/0.1/analytics-resources-impl.js [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+    if (!p || typeof p['getEntriesByType'] !== 'function') {
+      // ResourceTiming interface is not supported.
+      return null;
+    }
+    return p.getEntriesByType('resource');
+  }
+
+  /**
+   * Get DOM elements that match type and filter.
+   * @param {String} element type
```
Comment (dvoytenko) [inline extensions/amp-analytics-resources/0.1/analytics-resources-impl.js]: Lower-case `string`

File: spec/amp-var-substitutions.md [hunk 15]
```diff
@@ -96,6 +96,115 @@ For instance:
 ```
 may make a request to something like `https://foo.com/pixel?href=https%3A%2F%2Fpinterest.com%2F`.
 
+### DOM_EXT_IMG_COUNT
```
Comment (dvoytenko) [inline spec/amp-var-substitutions.md]: /cc @avimehta what's your advise on where to keep this documentation? Do we normally combine it all together in one doc or keep it within the extension itself and just link it out?

File: build-system/tasks/compile.js (hunk 6 above)
Comment (dvoytenko) [inline build-system/tasks/compile.js]: Also, @erwinmombay, what's our current 3rd-party code linking policy? Do we do it via npm or `third_party`?

**Comment by dvoytenko** [comment]
@querymetrics @avimehta Thanks! I left a few comments. Overall it looks good, but I'd really like to hear some performance metrics as well on how fast some of these features execute. I think, number wise, `img` lookup is the most controversial to me since I'm not sure if actually does what you wanted it to do.

**Comment by mrjoro** [comment]
Thank you for going through the effort of creating this PR.  It hasn't been updated in a while, so we're going to close it for now.  If you are still actively working on this PR or find yourself blocked on something/someone please reopen the PR with a comment giving the latest status.

**Pull request closed without merging**